t/var.lua

```diff
@@ -0,0 +1,229 @@
+-- Copyright (C) Yichun Zhang (agentzh)
+
+
+local ffi = require "ffi"
+local base = require "resty.core.base"
+
+
+local C = ffi.C
+local ffi_new = ffi.new
+local ffi_str = ffi.string
+local type = type
+local error = error
+local tostring = tostring
+local setmetatable = setmetatable
+local get_request = base.get_request
+local get_string_buf = base.get_string_buf
+local get_size_ptr = base.get_size_ptr
+local new_tab = base.new_tab
+local subsystem = ngx.config.subsystem
+
+
+local ngx_lua_ffi_var_get
+local ngx_lua_ffi_var_set
+
+local ngx_lua_kong_ffi_var_get_by_index
+local ngx_lua_kong_ffi_var_set_by_index
+local ngx_lua_kong_ffi_var_load_index
+
+
+local ERR_BUF_SIZE = 256
+
+local variable_index = {}
+
+-- Changed to distinguish between stock ngx.var
+ngx.var_indexed = new_tab(0, 0)
+
+
+if subsystem == "http" then
+    ffi.cdef[[
+    int ngx_http_lua_ffi_var_get(ngx_http_request_t *r,
+        const char *name_data, size_t name_len, char *lowcase_buf,
+        int capture_id, char **value, size_t *value_len, char **err);
+
+    int ngx_http_lua_ffi_var_set(ngx_http_request_t *r,
+        const unsigned char *name_data, size_t name_len,
+        unsigned char *lowcase_buf, const unsigned char *value,
+        size_t value_len, unsigned char *errbuf, size_t *errlen);
+    
+    int ngx_http_lua_kong_ffi_var_get_by_index(ngx_http_request_t *r,
+        unsigned int var_index, char **value, size_t *value_len, char **err);
+
+    int ngx_http_lua_kong_ffi_var_set_by_index(ngx_http_request_t *r,
+        unsigned int index, const unsigned char *value, size_t value_len,
+        char **err);
+
+    int ngx_http_lua_kong_ffi_var_load_index(const char *name_data,
+        size_t name_len, unsigned int *index, char **err);
+    ]]
+
+    ngx_lua_ffi_var_get = C.ngx_http_lua_ffi_var_get
+    ngx_lua_ffi_var_set = C.ngx_http_lua_ffi_var_set
+
+    ngx_lua_kong_ffi_var_get_by_index = C.ngx_http_lua_kong_ffi_var_get_by_index
+    ngx_lua_kong_ffi_var_set_by_index = C.ngx_http_lua_kong_ffi_var_set_by_index
+    ngx_lua_kong_ffi_var_load_index = C.ngx_http_lua_kong_ffi_var_load_index
+
+elseif subsystem == "stream" then
+    ffi.cdef[[
+    int ngx_stream_lua_ffi_var_get(ngx_stream_lua_request_t *r,
+        const char *name_data, size_t name_len, char *lowcase_buf,
+        int capture_id, char **value, size_t *value_len, char **err);
+
+    int ngx_stream_lua_ffi_var_set(ngx_stream_lua_request_t *r,
+        const unsigned char *name_data, size_t name_len,
+        unsigned char *lowcase_buf, const unsigned char *value,
+        size_t value_len, unsigned char *errbuf, size_t *errlen);
+    ]]
+
+    ngx_lua_ffi_var_get = C.ngx_stream_lua_ffi_var_get
+    ngx_lua_ffi_var_set = C.ngx_stream_lua_ffi_var_set
+end
+
+
+local value_ptr = ffi_new("unsigned char *[1]")
+local int_ptr = ffi_new("unsigned int [1]")
+local errmsg = base.get_errmsg_ptr()
+
+local function var_load_index(name)
```

Comment:
The index will be maintained by (patched) lua-resty-core and user won't notice a difference after this patch. Even we
return the actual index from `var_load_index`, it's more for debugging purpose, user are not allowed to get/set a
variable even with a index.
I think `ngx_http_get_variable_index` does a O(n) lookup and it could be expensive if we don't load it ahead of time.